app/src/test/java/org/simple/clinic/bp/entry/BloodPressureValidationMockDateValidatorTest.kt

```diff
@@ -0,0 +1,168 @@
+  @Test
+  @Suppress("IMPLICIT_CAST_TO_ANY")
+  @Parameters(method = "params for showing date validation errors")
+  fun `when save is clicked, date entry is active and input is invalid then validation errors should be shown`(
+      testParams: InvalidDateTestParams
+  ) {
+    val (openAs, day, month, year, errorResult, uiChangeVerification) = testParams
+
+    whenever(dateValidator.validate(any(), any())).doReturn(errorResult)
+    whenever(bloodPressureRepository.measurement(any())).doReturn(Observable.never())
+
+    uiEvents.run {
+      onNext(SheetCreated(openAs = openAs))
+      onNext(ScreenChanged(BloodPressureEntrySheet.ScreenType.DATE_ENTRY))
+      onNext(DayChanged(day))
+      onNext(MonthChanged(month))
+      onNext(YearChanged(year.takeLast(2)))
+      onNext(SaveClicked)
+    }
+
+    verify(bloodPressureRepository, never()).saveMeasurement(any(), any(), any(), any(), any(), any())
+    verify(bloodPressureRepository, never()).updateMeasurement(any())
+    verify(ui, never()).setBpSavedResultAndFinish()
+
+    uiChangeVerification(ui)
+  }
+
+  @Suppress("unused")
+  fun `params for showing date validation errors`(): List<InvalidDateTestParams> {
+    val existingBpUuid = UUID.fromString("2c4eccbb-d1bc-4c7c-b1ec-60a13acfeea4")
+    return listOf(
+        InvalidDateTestParams(OpenAs.New(patientUuid), "01", "01", "2099", DateIsInFuture) { ui: Ui -> verify(ui).showDateIsInFutureError() },
+        InvalidDateTestParams(OpenAs.Update(existingBpUuid), "01", "01", "2099", DateIsInFuture) { ui: Ui -> verify(ui).showDateIsInFutureError() })
+  }
+
+  @Test
+  @Suppress("IMPLICIT_CAST_TO_ANY")
+  @Parameters(method = "params for checking valid date input")
+  fun `when save is clicked, date entry is active, but input is invalid then BP measurement should not be saved`(
+      testParams: DoNotSaveBpWithInvalidDateTestParams
+  ) {
+    val (openAs, result) = testParams
+
+    whenever(dateValidator.validate(any(), any())).doReturn(result)
+    whenever(bloodPressureRepository.measurement(any())).doReturn(Observable.never())
+
+    uiEvents.onNext(SheetCreated(openAs))
+    uiEvents.onNext(ScreenChanged(BloodPressureEntrySheet.ScreenType.DATE_ENTRY))
+    uiEvents.onNext(DayChanged("1"))
+    uiEvents.onNext(MonthChanged("4"))
+    uiEvents.onNext(YearChanged("9"))
+    uiEvents.onNext(SaveClicked)
+
+    when (openAs) {
+      is OpenAs.New -> verify(bloodPressureRepository, never()).saveMeasurement(any(), any(), any(), any(), any(), any())
+      is OpenAs.Update -> verify(bloodPressureRepository, never()).updateMeasurement(any())
+      else -> throw AssertionError()
+    }
+
+    verify(ui, never()).setBpSavedResultAndFinish()
+    verify(dateValidator).validate("01/04/1909")
+  }
+
+  @Suppress("Unused")
+  private fun `params for checking valid date input`(): List<DoNotSaveBpWithInvalidDateTestParams> {
+    return listOf(
+        DoNotSaveBpWithInvalidDateTestParams(OpenAs.New(patientUuid), InvalidPattern),
```

Comment:
Isn't the `InvalidPattern` case tested in `BloodPressureEntrySheetControllerTest`?

Reply:
Removed.